### PR TITLE
coverity: iio_stresstest : fix Untrusted value as argument (TAINTED_SCALAR)

### PR DESCRIPTION
**Pull request by rgetz**
According to a few places I found, the passing an argc to strtoul
without checking it's length first, will cause this error to be flagged,
so check the length.

Hopefully this finally resolves things.

Signed-off-by: Robin Getz <robin.getz@analog.com>

### CODE REVIEW

File: tests/iio_stresstest.c [hunk 1]
```diff
@@ -82,11 +89,14 @@ static size_t cache_line_size(void)
...
 #elif __linux__
 	FILE * p = 0;
+	int ret;
 
 	p = fopen("/sys/devices/system/cpu/cpu0/cache/index0/coherency_line_size", "r");
 	if (p) {
-		fscanf(p, "%zu", &cacheline);
+		ret = fscanf(p, "%zu", &cacheline);
 		fclose(p);
+		if (!ret || ret == EOF)
```
Comment (commodo) [inline tests/iio_stresstest.c:98]: maybe coverity would complain that negative values are checked?
no idea;

in any case, maybe let's go for the simplest check for `ret = fscanf()`; i.e. `if (ret != 1) cacheline = 0`;
we know that the function should return exactly 1 scanned element;
enforcing anything-but 1 scanned elem returned should cover all other returns;

File: tests/iio_stresstest.c [hunk 2]
```diff
@@ -388,18 +398,37 @@ static void *client_thread(void *data)
 	return (void *)EXIT_FAILURE;
 }
 
-static unsigned long int clamp(const char *name, unsigned long int val,
+static unsigned long int sanitize_clamp(const char *name, const char *val,
```
Comment (commodo) [inline tests/iio_stresstest.c:401]: if we rename `const char *val` -> `const char *argv`, we can rename `tmp` -> `val` as a local variable;
that would reduce some bits of patch in the clamp() part;

File: tests/iio_stresstest.c [hunk 3]
```diff
@@ -388,18 +398,37 @@ static void *client_thread(void *data)
...
 		unsigned long int min, unsigned long int max)
 {
-	if (val > max) {
-		val = max;
-		fprintf(stderr, "Clamped %s to max %lu", name, max);
+
+	unsigned long int tmp;
+	char buf[20];
+
+	if(!val) {
```
Comment (commodo) [inline tests/iio_stresstest.c:408]: nitpick: `if__space__() {`

File: tests/iio_stresstest.c [hunk 4]
```diff
@@ -388,18 +398,37 @@ static void *client_thread(void *data)
...
 }
 
-static unsigned long int clamp(const char *name, unsigned long int val,
+static unsigned long int sanitize_clamp(const char *name, const char *val,
 		unsigned long int min, unsigned long int max)
 {
-	if (val > max) {
-		val = max;
-		fprintf(stderr, "Clamped %s to max %lu", name, max);
+
```
Comment (commodo) [inline tests/iio_stresstest.c:404]: nitpick: extra line

File: tests/iio_stresstest.c [hunk 5]
```diff
@@ -388,18 +398,37 @@ static void *client_thread(void *data)
...
+	if(!val) {
+		tmp = 0;
+	} else {
+
+		/* find the length of chars of max, and make sure the string
+		 * is shorter than that
+		 */
+		iio_snprintf(buf, sizeof(buf), "%lu", max);
+		if (strnlen(val, strlen(buf) + 1) <= strlen(buf)) /* Flawfinder: ignore */
```
Comment (commodo) [inline tests/iio_stresstest.c:416]: umm, i feel this is where pleasing coverity starts to reduce the value of the code [in terms of readability, maintainability,  understanding the intent of the code, etc];
[ to-be-fair: i also [sometimes] fall easily into the trap of pleasing static-checkers ; ]

using a buffer for argv[idx]  + strlen()/strlnlen() doesn't seem to be more efficient than simply using strtol()
if there is no null terminator, things will go sour anyway;
i'd expect strltoul() to use strlen() or something equivalent anyway;

now, there could be some better way to please coverity and still make this readable;
but if there is a flag/comment for coverity to ignore this, it's probably not a bad idea [either];

[fwiw: the idea of parsing argv[] in this function is nice though];

File: tests/iio_stresstest.c (hunk 4 above)
Comment (rgetz) [reply tests/iio_stresstest.c:404]: fixed

File: tests/iio_stresstest.c (hunk 3 above)
Comment (rgetz) [reply tests/iio_stresstest.c]: fixed

File: tests/iio_stresstest.c (hunk 2 above)
Comment (rgetz) [reply tests/iio_stresstest.c]: done

File: tests/iio_stresstest.c (hunk 1 above)
Comment (rgetz) [reply tests/iio_stresstest.c]: updated

File: tests/iio_stresstest.c (hunk 5 above)
Comment (rgetz) [reply tests/iio_stresstest.c]: Updated with a simpler solution.

File: tests/iio_stresstest.c [hunk 6]
```diff
@@ -388,16 +398,28 @@ static void *client_thread(void *data)
...
 {
+
+	unsigned long int val;
+	char buf[20];
+
+	if (!argv) {
+		val = 0;
+	} else {
+		/* sanitized buffer by taking first 20 (or less) char */
+		iio_snprintf(buf, sizeof(buf), "%s",argv);
```
Comment (commodo) [inline tests/iio_stresstest.c:412]: nitpick: `iio_snprintf(buf, sizeof(buf), "%s",__space__argv);`

Comment (rgetz) [reply tests/iio_stresstest.c]: fixed - sorry - need to get the formatter installed...

**Comment by rgetz** [comment]
Finally!

image

**Pull request merged**